contrib/internal/httptrace/inferred_proxy.go

```diff
@@ -0,0 +1,138 @@
+	ProxyHeaderStage       = "X-Dd-Proxy-Stage"
+)
+
+type ProxyDetails struct {
+	SpanName  string `json:"spanName"`
+	Component string `json:"component"`
+}
+
+var (
+	supportedProxies = map[string]ProxyDetails{
+		"aws-apigateway": {
+			SpanName:  "aws.apigateway",
+			Component: "aws-apigateway",
+		},
+	}
+)
+
+type ProxyContext struct {
+	RequestTime     string `json:"requestTime"`
+	Method          string `json:"method"`
+	Path            string `json:"path"`
+	Stage           string `json:"stage"`
+	DomainName      string `json:"domainName"`
+	ProxySystemName string `json:"proxySystemName"`
+}
+
+func extractInferredProxyContext(headers http.Header) *ProxyContext {
+	_, exists := headers[ProxyHeaderStartTimeMs]
+	if !exists {
+		log.Debug("Proxy header start time does not exist")
+		return nil
+	}
+
+	proxyHeaderSystem, exists := headers[ProxyHeaderSystem]
+	if !exists {
+		log.Debug("Proxy header system does not exist")
+		return nil
+	}
+
+	if _, ok := supportedProxies[proxyHeaderSystem[0]]; !ok {
+		log.Debug("Unsupported Proxy header system")
+		return nil
+	}
+
+	return &ProxyContext{
+		RequestTime:     headers[ProxyHeaderStartTimeMs][0],
+		Method:          headers[ProxyHeaderHttpMethod][0],
+		Path:            headers[ProxyHeaderPath][0],
+		Stage:           headers[ProxyHeaderStage][0],
+		DomainName:      headers[ProxyHeaderDomain][0],
+		ProxySystemName: headers[ProxyHeaderSystem][0],
+	}
+
+}
+
+func tryCreateInferredProxySpan(headers http.Header, parent ddtrace.SpanContext, opts ...StartSpanOption) tracer.Span {
+	if headers == nil {
+		log.Debug("Headers do not exist")
+		return nil
+
+	}
+
+	requestProxyContext := extractInferredProxyContext(headers)
+	if requestProxyContext == nil {
+		log.Debug("Unable to extract inferred proxy context")
+		return nil
+	}
+
+	proxySpanInfo := supportedProxies[requestProxyContext.ProxySystemName]
+	log.Debug(`Successfully extracted inferred span info ${proxyContext} for proxy: ${proxyContext.proxySystemName}`)
+
+	startTimeUnixMilli, err := strconv.ParseInt(requestProxyContext.RequestTime, 10, 64)
+	if err != nil {
+		log.Debug("Error parsing time string: %v", err)
+		return nil
+	}
+	startTime := time.UnixMilli(startTimeUnixMilli)
+
+	configService := requestProxyContext.DomainName
+	if configService == "" {
+		configService = globalconfig.ServiceName()
+	}
+
+	optsLocal := make([]StartSpanOption, len(opts), len(opts)+1)
+	copy(optsLocal, opts)
+
+	optsLocal = append(optsLocal,
+		func(cfg *ddtrace.StartSpanConfig) {
+			if cfg.Tags == nil {
+				cfg.Tags = make(map[string]interface{})
+			}
+
+			cfg.Parent = parent
+			cfg.StartTime = startTime
+
+			cfg.Tags[ext.SpanType] = ext.SpanTypeWeb
+			cfg.Tags[ext.ServiceName] = configService
+			cfg.Tags[ext.Component] = proxySpanInfo.Component
+			cfg.Tags[ext.HTTPMethod] = requestProxyContext.Method
+			cfg.Tags[ext.HTTPURL] = requestProxyContext.DomainName + requestProxyContext.Path
+			cfg.Tags[ext.HTTPRoute] = requestProxyContext.Path
```

Comment:
`http.route` shouldn't be used unless it is known for sure it is the low cardinality route set by the framework / library ([otel docs](https://opentelemetry.io/docs/specs/semconv/attributes-registry/http/#http-route))

> [5] http.route: MUST NOT be populated when this is not supported by the HTTP server framework as the route attribute should have low-cardinality and the URI path can NOT substitute it. SHOULD include the [application root](https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-server-definitions) if there is one.

I suspect since API gateway (or any other proxy) won't have the routing information from the framework (e.g. that there was a route `/users/:id` registered) and this will be instead something like `/users/123`. Could you please verify this? 🙏

Reply:
Removed http.route until we get confirmation